srml/support/procedural/src/construct_runtime/parse.rs

```diff
@@ -0,0 +1,211 @@
+	syn::custom_keyword!(NodeBlock);
+	syn::custom_keyword!(UncheckedExtrinsic);
+}
+
+#[derive(Parse, ToTokens, Debug)]
+pub struct RuntimeDefinition {
+	pub visibility_token: Token![pub],
+	pub enum_token: Token![enum],
+	pub name: Ident,
+	pub where_section: WhereSection,
+	pub modules: ext::Braces<ext::Punctuated<ModuleDeclaration, Token![,]>>,
+}
+
+#[derive(Parse, ToTokens, Debug)]
+pub struct WhereSection {
+	pub token: Token![where],
+	pub block_token: keyword::Block,
+	pub block_eq: Token![=],
+	pub block: syn::TypePath,
+	pub block_sep: Token![,],
+	pub node_block_token: keyword::NodeBlock,
+	pub node_block_eq: Token![=],
+	pub node_block: syn::TypePath,
+	pub node_block_sep: Token![,],
+	pub unchecked_extrinsic_token: keyword::UncheckedExtrinsic,
+	pub unchecked_extrinsic_eq: Token![=],
+	pub unchecked_extrinsic: syn::TypePath,
+	pub trailing_comme: ext::Opt<Token![,]>,
+}
+
+#[derive(Parse, ToTokens, Debug)]
+pub struct ModuleDeclaration {
+	pub name: Ident,
+	pub name_colon: Token![:],
+	pub module: Ident,
+	pub instance: ModuleInstanceWrapper,
+	pub details: ModuleDetailsWrapper,
+}
+
+#[derive(ToTokens, Debug)]
+pub struct ModuleInstanceWrapper {
+	pub inner: Option<ModuleInstance>,
+}
+
+impl Parse for ModuleInstanceWrapper {
+	fn parse(input: ParseStream) -> Result<Self> {
+		// In this case we're sure it needs to be a ModuleInstance
+		if input.peek(Token![::]) && input.peek3(Token![<]) {
+			let inner = Some(input.parse()?);
+			Ok(ModuleInstanceWrapper { inner })
+		} else {
+			Ok(ModuleInstanceWrapper { inner: None })
+		}
+	}
+}
+
+#[derive(ToTokens, Debug)]
+pub struct ModuleDetailsWrapper {
+	pub inner: Option<ModuleDetails>,
+}
+
+impl Parse for ModuleDetailsWrapper {
+	fn parse(input: ParseStream) -> Result<Self> {
+		// In this case we're sure it needs to be a ModuleDetails
+		if input.peek(Token![::]) {
+			let inner = Some(input.parse()?);
+			Ok(ModuleDetailsWrapper { inner })
+		} else {
+			Ok(ModuleDetailsWrapper { inner: None })
+		}
+	}
+}
+
+#[derive(Parse, ToTokens, Debug)]
+pub struct ModuleInstance {
+	pub colons: Token![::],
+	pub lt: Token![<],
+	pub name: Ident,
+	pub gt: Token![>],
+}
+
+#[derive(Parse, ToTokens, Debug)]
+pub struct ModuleDetails {
+	pub colons: Token![::],
+	pub entries: ext::Braces<ext::Punctuated<ModuleEntryWrapper, Token![,]>>,
+}
+
+#[derive(ToTokens, Debug)]
+pub struct ModuleEntryWrapper {
+	pub inner: ModuleEntry,
+}
+
+impl Parse for ModuleEntryWrapper {
+	fn parse(input: ParseStream) -> Result<Self> {
+		ModuleEntry::parse(input)
+			.map(|inner| ModuleEntryWrapper { inner })
+			.map_err(|_| input.error("Expected `default` or module export name (e.g. Call, Event, etc.)"))
+	}
+}
+
+#[derive(Parse, ToTokens, Debug)]
+pub enum ModuleEntry {
+	Default(Token![default]),
+	Part(ModulePart),
+}
+
+#[derive(Parse, ToTokens, Debug, Clone)]
+pub struct ModulePart {
+	pub name: Ident,
+	pub generics: syn::Generics,
+	pub args: ext::Opt<ext::Parens<ext::Punctuated<Ident, Token![,]>>>,
```

Comment:
I'm not sure this option is legit as well, if the inner thing is not a valid ident for example then the whole option will be None and the error will be on the invalidity of ModuleEntryWrapper I think

Reply:
also only inherent have this args, all other must be checked to have none